pkg/kn/commands/plugin/plugin_verifier.go

```diff
@@ -0,0 +1,259 @@
+	return &pluginVerifier{
+		root:        root,
+		seenPlugins: make(map[string]string),
+	}
+}
+
+// permission bits for execute
+const (
+	UserExecute  = 1 << 6
+	GroupExecute = 1 << 3
+	OtherExecute = 1 << 0
+)
+
+// Verify implements pathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+// This method is not idempotent and must be called for each path only once.
+func (v *pluginVerifier) verify(eaw errorsAndWarnings, path string) errorsAndWarnings {
+	if v.root == nil {
+		return eaw.addError("unable to verify path with nil root")
+	}
+
+	// Verify that plugin actually exists
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		if err == os.ErrNotExist {
+			return eaw.addError("cannot find plugin in %s", path)
+		}
+		return eaw.addError("cannot stat %s: %v", path, err)
+	}
+
+	eaw = v.addErrorIfWrongPrefix(eaw, path)
+	eaw = v.addWarningIfNotExecutable(eaw, path, fileInfo)
+	eaw = v.addWarningIfAlreadySeen(eaw, path)
+	eaw = v.addErrorIfOverwritingExistingCommand(eaw, path)
+
+	// Remember each verified plugin for duplicate check
+	v.seenPlugins[filepath.Base(path)] = path
+
+	return eaw
+}
+
+func (v *pluginVerifier) addWarningIfAlreadySeen(eaw errorsAndWarnings, path string) errorsAndWarnings {
+	fileName := filepath.Base(path)
+	if existingPath, ok := v.seenPlugins[fileName]; ok {
+		return eaw.addWarning("%s is ignored because it is shadowed by a equally named plugin: %s.", path, existingPath)
+	}
+	return eaw
+}
+
+func (v *pluginVerifier) addErrorIfOverwritingExistingCommand(eaw errorsAndWarnings, path string) errorsAndWarnings {
+	fileName := filepath.Base(path)
+	cmds := strings.Split(fileName, "-")
+	if len(cmds) < 2 {
+		return eaw.addError("%s is not a valid plugin filename as its missing a prefix", fileName)
+	}
+	cmds = cmds[1:]
+
+	// Check both, commands with underscore and with dash because plugins can be called with both
+	overwrittenCommands := make(map[string]bool)
+	for _, c := range [][]string{cmds, convertUnderscoresToDashes(cmds)} {
+		cmd, _, err := v.root.Find(c)
+		if err == nil {
+			overwrittenCommands[cmd.CommandPath()] = true
+		}
+	}
+	for command := range overwrittenCommands {
+		eaw.addError("%s overwrites existing built-in command: %s", fileName, command)
+	}
+	return eaw
+}
+
+func (v *pluginVerifier) addErrorIfWrongPrefix(eaw errorsAndWarnings, path string) errorsAndWarnings {
+	fileName := filepath.Base(path)
+	// Only pick the first prefix as it is very like that it will be reduced to
+	// a single prefix anyway (PR pending)
+	prefix := ValidPluginFilenamePrefixes[0]
+	if !strings.HasPrefix(fileName, prefix+"-") {
+		eaw.addWarning("%s plugin doesn't start with plugin prefix %s", fileName, prefix)
+	}
+	return eaw
+}
+
+func (v *pluginVerifier) addWarningIfNotExecutable(eaw errorsAndWarnings, path string, fileInfo os.FileInfo) errorsAndWarnings {
```

Comment:
Would it make sense to divide this into smaller helper func for each of the different cases?  This func is close to 70 lines long...